src/new/wallet.js

```diff
@@ -1171,202 +1055,50 @@ class HathorWallet extends EventEmitter {
-    this.setPreProcessedData('balanceByToken', tokensBalance);
-  }
-
-  /**
-   * Set data in the pre processed object
-   *
-   * @param {string} key Key of the pre processed object to be added
-   * @param {Any} value Value of the pre processed object to be added
-   *
-   * @memberof HathorWallet
-   * @inner
-   **/
-  setPreProcessedData(key, value) {
-    this.preProcessedData[key] = value;
-  }
-
-  /**
-   * Get data in the pre processed object
-   * If pre processed data is empty, we generate it and return
-   *
-   * @param {string} key Key of the pre processed object to get the value
-   *
-   * @return {Any} Value of the pre processed object
-   *
-   * @memberof HathorWallet
-   * @inner
-   **/
-  getPreProcessedData(key) {
-    if (Object.keys(this.preProcessedData).length === 0) {
-      throw new Error('Wallet data has not been processed yet');
-    }
-    return this.preProcessedData[key];
-  }
-
   /**
    * Call the method to process data and resume with the correct state after processing.
    *
    * @returns {Promise} A promise that resolves when the wallet is done processing the tx queue.
    */
   async onEnterStateProcessing() {
     // Started processing state now, so we prepare the local data to support using this facade interchangable with wallet service facade in both wallets
-    return this.preProcessWalletData()
+    return this.storage.processHistory()
       .then(() => { this.setState(HathorWallet.READY); })
       .catch(() => { this.setState(HathorWallet.ERROR); });
   }
 
   setState(state) {
     if (state === HathorWallet.PROCESSING && state !== this.state) {
+      // XXX: will not await this so we can process history on background.
       this.onEnterStateProcessing();
     }
     this.state = state;
     this.emit('state', state);
   }
 
-  onNewTx(wsData) {
-    storage.setStore(this.store);
+  async onNewTx(wsData) {
     const newTx = wsData.history;
+    const storageTx = await this.storage.getTx(newTx.tx_id);
+    const isNewTx = storageTx === null;
 
-    const walletData = wallet.getWalletData();
-    const historyTransactions = 'historyTransactions' in walletData ? walletData.historyTransactions : {};
-    const allTokens = 'allTokens' in walletData ? walletData.allTokens : [];
+    // Save the transaction in the storage
+    await this.storage.addTx(newTx);
 
-    let isNewTx = true;
-    if (newTx.tx_id in historyTransactions) {
-      isNewTx = false;
+    // check gap-limit and load more addresses if needed
+    const fillGapLimit = await checkGapLimit(this.storage);
+    if (fillGapLimit !== null) {
+      await syncHistory(fillGapLimit.nextIndex, fillGapLimit.count, this.storage, this.conn);
     }
-
-    wallet.updateHistoryData(historyTransactions, allTokens, [newTx], null, walletData, null, this.conn, this.store);
-
-    this.onTxArrived(newTx, isNewTx);
+    // Process history to update metadatas
+    await this.storage.processHistory();
```

Comment:
Do we need to process the full history in every new transaction?

Reply:
Unfortunately yes, but in some cases this can actually be better than the pre-processing we had.

The old method had a list of transactions sorted by timestamp for each token and we would push any new transaction to the end then sort the array (effectively bubbling the last transaction to the front).
This operation should be `O(n*log(n))` and since most wallets only have HTR it would not help to split in tokens.

The processHistory is `O(n)` and while it has more overhead for smaller wallets it should not be noticeable, but it definetely will have a positive impact on wallets with thousands of transactions.

This behavior is still not ideal but we cannot escape this until we resolve the issue with repeated transaction events (possibly with the reliable integrations)

Reply:
The issue #479  was created to track this discussion